### DISCUSSION TITLE
MauiApplicationInsights questions

### DISCUSSION THREAD

**Discussion by vovanb** [VladislavAntonyuk/MauiSamples#440] [Q&A] [answered]
Hi , I have two questions:
1. For some reason I can not found in my Application Insights in Azure portal any logs/exceptions. Can you please point how exactly I can access results you showed [here ](https://ik.imagekit.io/VladislavAntonyuk/vladislavantonyuk/articles/46/appinsights.png) .How much time should pass from event that it will appear in portal?
2. How I can use Application Insights from Android MainActivity.cs?

**Comment by VladislavAntonyuk** [comment]
Hello,
1. Check your connection string and make sure your device is connected to the Internet. Also make sure you registered Application Insights in Program.cs (you can put your connection string in my sample app and check if you see logs)
2. Application Insights is a provider for the log outputs. So you only need to use Logger in your class (MainActivity) and logs will be automatically written to Azure.

**Reply by vovanb** [reply]
@VladislavAntonyuk 
thanks for answers. 
for #1 -  the device was not connected  to Internet  - is it possible to get logs after internet connection restored?
for #2 - I do not have Logger in class MainActivity : MauiAppCompatActivity class? Could you please explain ?

**Reply by VladislavAntonyuk** [reply, marked as answer]
1. It's a good question about how Application Insights works when a device is offline. As far as I know, it collects logs and you can force it to send logs when a device connects to the internet. The alternative to Microsoft Logger, you can use Serilog, which has a method Flush. By calling this method it writes log to sinks (ApplicationInsights in your case)
2. I can access it from the DI. `var logger = Application.Current.Handler.MauiContext.Services.GetService<ILogger>();` (https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/dependency-injection?view=net-maui-8.0)

**Reply by vovanb** [reply]
For me it is not working .... Application.Current gives error:
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	CS1061	'Application' does not contain a definition for 'Current' and no accessible extension method 'Current' accepting a first argument of type 'Application' could be found (are you missing a using directive or an assembly reference?)	MauiApplicationInsights (net8.0-android)	main\MauiApplicationInsights\Platforms\Android\MainActivity.cs	15	Active	
```

changed it to 
```
        public MainActivity()
	{
		//var logger = Application.Current.Handler.MauiContext.Services.GetService<ILogger>();

		var logger = App.Current?.Handler?.MauiContext?.Services.GetService<ILogger>();
		logger?.LogInformation("MainActivity {0}", DateTime.Now);

	}
```

but logger I got is null
I am using your example as is 

`

**Reply by vovanb** [reply]
found a solution :
```
protected override  void OnCreate(Bundle? savedInstanceState)
{
	base.OnCreate(savedInstanceState);

	var logger = (Application as IPlatformApplication)?.Services.GetService<ILogger<MainActivity>>();
	logger?.LogInformation("Hello World");

}
```

**Reply by vovanb** [reply]
please also note that if you press fast to number above 6 and app crashed the exception `logger.LogError(exception, "Something went wrong: {Count}", count); ` did not write into ApplicationInsights

**Reply by VladislavAntonyuk** [reply]
The code I sent you works. Application is MauiApplication, but you loaded Android Application. Change the namespace.
There is a delay up to 5 minutes before logs appear in Application Insights.
Also it may send logs if you start app again.

**Reply by vovanb** [reply]
The exception "**Count > 5**" does not appear after I restart application. The two exceptions "**Count > 2**"  do appear. I believe you can easily reproduce it with your sample without any changes

**Reply by vovanb** [reply]
found solution - you can update your example:
in MauiProgram:
`  public static InMemoryChannel channel = new InMemoryChannel();`
then 
```
 builder.Logging.AddApplicationInsights(configuration =>
 {
     configuration.TelemetryChannel = channel; 
     configuration.TelemetryInitializers.Add(new ApplicationInitializer());
 
```
and then in global exception handler or in OnStopped (for some reason OnDestroying never called)
`   MauiProgram.channel.Flush();
`

**Reply by vovanb** [reply]
> The code I sent you works. Application is MauiApplication, but you loaded Android Application. Change the namespace. There is a delay up to 5 minutes before logs appear in Application Insights. Also it may send logs if you start app again.

It is always null
